semantics metrics assertion, part 1: auxiliary GomegaMatcher-s

Ray-Eldath opened the pull request:
(feat) GomegaMatcher for Array & Slice: AllPassed and AnyFailed, tested & documented, in package matcher. **Go for `container_test.go` to see how to use these two convienent matcher when working with array or slice.**
(feat) GomegaMatcher for metrics checking with PromQL, documented, in package matcher.
(feat) GomegaMatcher for panics with message printed (and an omissible parameter "name"), documented, in package matcher.
(feat) ConsistentlyWithMetrics for another set of defaults, tested, in package assertion.
(refactor) refactor ParameterizedTestCase with Golang builtin parsing method.
(refactor) refactor demonstrating tikv_test.go with newly added PromQL matcher.

Ray-Eldath committed:
(feat) GomegaMatcher for Array & Slice: AllShouldPassed and AnyShouldFailed, tested.
(feat) GomegaMatcher for metrics checking with PromQL.
(feat) GomegaMatcher for panics with message printed (and an omissible parameter "name").
(refactor) refactor demonstrating tikv_test.go with newly added PromQL matcher.

cosven commented on pkg/ginkgo_helper/metrics.go:
```diff
@@ -0,0 +1,113 @@
+)
+
+func BuildPrometheusAPI(tc resource.TiDBCluster) (v1.API, error) {
+	url, err := tc.ServiceURL(resource.Prometheus)
```
I think we should not import infra sdk in other pacakges.

Ray-Eldath replied:
We could let user pass in the result of `ServiceURL` directly.

cosven replied:
Yeah!

cosven commented on pkg/ginkgo_helper/metrics.go:
```diff
@@ -0,0 +1,113 @@
+	gomega.SetDefaultConsistentlyDuration(3 * time.Hour)
+	gomega.SetDefaultConsistentlyPollingInterval(1 * time.Minute)
+	gomega.SetDefaultEventuallyTimeout(3 * time.Hour)
+	gomega.SetDefaultEventuallyPollingInterval(1 * time.Minute)
```
Why change default values here?

Ray-Eldath replied:
default value is *"`Eventually` will poll every 10 milliseconds for up to 1 second and `Consistently` will monitor every 10 milliseconds for up to 100 milliseconds"* which is undoubtedly too frequent for a metrics checking, and the timeout is too short as we need to moniter the status of a cluster in a relatively long time as well.

cosven replied:
Can we only set default value for metrics-checking? It seems that this PR changes the  global default value.

Ray-Eldath replied:
seems not supportted by gomega. Or we could make our own version of `Consistently` (maybe `ConsistentlyOnMetrics`...?) that works as a pure delegate of the original `Consistently` with longer interval & duration passed in.

cosven replied:
Maybe we can define some varaibles and case can use them in metrics matcher?

cosven commented on testcase/oom/tikv_test.go:
```diff
@@ -11,12 +11,12 @@ import (
 
-	qe_metrics "github.com/PingCAP-QE/metrics-checker/pkg/metrics"
 	_ "github.com/go-sql-driver/mysql"
+	infraresource "github.com/pingcap/test-infra/sdk/resource"
```
IMO, `infra_resource` has better readability 🤔.

Ray-Eldath replied:
but will makes GoLand unhappy 😂😂

cosven replied:
I see😂. I checked the blog [package-names](https://blog.golang.org/package-names).

Maybe we can use name `resource` directly?

cosven commented on pkg/ginkgo_helper/metrics.go:
```diff
@@ -0,0 +1,113 @@
+	}
+}
+
+type promQLEvaluatedMatcher struct {
```
Can we seperate the gomega logic from promQL check logic?

Ray-Eldath replied:
the main PromQL logic is only a small function, roughly 15 lines of code (`pkg/ginkgo_helper/metrics.go:74`). I think keep the function inplace could make it more compact and more readable.

cosven replied:
LGTM.

cosven commented on pkg/ginkgo_helper/collection.go:
```diff
@@ -0,0 +1,140 @@
+	successPositions []int
+}
+
+func AnyShouldFailed(assertion types.GomegaMatcher) *anyShouldNotMatcher {
```
What does `AnyShouldFailed` mean?

cosven commented on pkg/ginkgo_helper/collection.go:
```diff
@@ -0,0 +1,140 @@
+
+type allShouldMatcher struct {
+	assertion types.GomegaMatcher
+	evaluated bool
```
It seems  the attribute `evaluated` is not used.

Ray-Eldath commented:
@cosven problems addressed, PTAL.

Ray-Eldath committed:
(feat) GomegaMatcher for Array & Slice: AllPassed and AnyFailed, tested & documented, in package matcher.
(feat) GomegaMatcher for metrics checking with PromQL, documented, in package matcher.
(feat) GomegaMatcher for panics with message printed (and an omissible parameter "name"), documented, in package matcher.
(feat) ConsistentlyWithMetrics for another set of defaults, tested, in package assertion.
(refactor) refactor ParameterizedTestCase with Golang builtin parsing method.
(refactor) refactor demonstrating tikv_test.go with newly added PromQL matcher.

Ray-Eldath commented:
@cosven now its completely dual, ptal. hope this is the last shot, then we can work on `endless`. 😉

cosven commented on pkg/ginkgo_helper/metrics.go:
```diff
@@ -0,0 +1,108 @@
+	"github.com/prometheus/common/model"
+)
+
+func BuildPrometheusAPI(prometheusURL url.URL) (v1.API, error) {
```
How about change the name to `NewPrometheusClient`?

cosven commented on pkg/ginkgo_helper/panic.go:
```diff
@@ -0,0 +1,50 @@
+	return p.delegate.NegatedFailureMessage(actual)
+}
+
+func PanicWithOutput(name ...string) types.GomegaMatcher {
```
We should add some docs for these dsl lator.

Ray-Eldath committed:
(misc) EditorConfig for better enforcing code style.

Ray-Eldath commented:
@cosven all done. :wink:

Ray-Eldath committed:
(*) merge & rename pkg.

cosven reviewed (approved):
LGTM
